File: golang/statistics/save_statistics.go
package statistics

import (
	"bench-pg-go/model/domain"
	"bench-pg-go/model/metrics"
	"bench-pg-go/statistics/model"
	"log"

	"github.com/samber/lo"
	"gonum.org/v1/gonum/floats"
	"gonum.org/v1/gonum/stat"
	"gorm.io/gorm"
)

type statCondition struct {
	Action      string
	FilterFn    func(metrics.DomainMetric) bool
	CalcPeriods bool
}

func newStatCondition(
	action string,
	filterFn func(metrics.DomainMetric) bool,
	calcPeriods bool,
) statCondition {
	return statCondition{Action: action, FilterFn: filterFn, CalcPeriods: calcPeriods}
}

var PROCESSING_RULES = []statCondition{
	newStatCondition("PROCESS_YEAR", filterByProcessYear, false),
	newStatCondition("PROCESS_PERIOD", filterByProcessPeriod, false),
	newStatCondition("YEAR_REPORT", filterByYearReport, true),
	newStatCondition("PERIOD_REPORT", filterByPeriodReport, true),
	newStatCondition("OPEN_PERIOD", filterByOpenPeriod, false),
	newStatCondition("CLOSE_PERIOD", filterByClosePeriod, false),
	newStatCondition("COST", filterByCost, true),
	newStatCondition("SALE", filterBySale, true),
	newStatCondition("PURCHASE", filterByPurchase, true),
}

func saveStatistics(db *gorm.DB, benchmarkId int64, domainMetrics []metrics.DomainMetric, startYear int, years int) {
	for _, rule := range PROCESSING_RULES {
		actionMetrics := lo.Filter(domainMetrics, func(item metrics.DomainMetric, index int) bool {
			return rule.FilterFn(item)
		})
		durations := lo.Map(actionMetrics, func(item metrics.DomainMetric, index int) float64 {
			return item.Duration
		})

		totalStat := calculateStatistics(durations)
		saveStat(db, benchmarkId, nil, nil, rule.Action, totalStat)

		for year := startYear; year < startYear+years; year++ {
			yearMetrics := lo.Filter(actionMetrics, func(item metrics.DomainMetric, index int) bool {
				return item.Year == year
			})
			durations := lo.Map(yearMetrics, func(item metrics.DomainMetric, index int) float64 {
				return item.Duration
			})
			stat := calculateStatistics(durations)
			saveStat(db, benchmarkId, &year, nil, rule.Action, stat)

			if rule.CalcPeriods {
				for month := 1; month <= 12; month++ {
					monthMetrics := lo.Filter(yearMetrics, func(item metrics.DomainMetric, index int) bool {
						return item.Period != nil && *item.Period == month
					})
					durations := lo.Map(monthMetrics, func(item metrics.DomainMetric, index int) float64 {
						return item.Duration
					})
					stat := calculateStatistics(durations)
					saveStat(db, benchmarkId, &year, &month, rule.Action, stat)
				}
			}

		}

	}
}

func saveStat(db *gorm.DB, benchmarkId int64, year *int, month *int, action string, s metrics.Statistics) {
	entity := model.StatisticEntity{
		Id:            0,
		BenchmarkId:   benchmarkId,
		Action:        action,
		Year:          year,
		Month:         month,
		TotalCount:    int64(s.TotalCount),
		TotalDuration: s.TotalDuration,
		OpsPerSec:     s.OpsPerSec,
		Min:           s.Min,
		P50:           s.P50,
		P75:           s.P75,
		P95:           s.P95,
		P99:           s.P99,
		P99_9:         s.P999,
		Max:           s.Max,
		Mean:          s.Mean,
		StdDev:        s.StdDev,
	}
	res := db.Create(&entity)
	if err := res.Error; err != nil {
		log.Fatalf("Error inserting metric: %v", err)
	}
}

func calculateStatistics(durations []float64) metrics.Statistics {
	totalCount := len(durations)
	totalDuration := floats.Sum(durations)
	var smin float64
	var p50 float64
	var p75 float64
	var p95 float64
	var p99 float64
	var p999 float64
	var smax float64
	var mean float64
	var stdDev float64
	if totalCount > 0 {
		stat.SortWeighted(durations, nil)
		smin = floats.Min(durations)
		p50 = stat.Quantile(0.5, stat.Empirical, durations, nil)
		p75 = stat.Quantile(0.75, stat.Empirical, durations, nil)
		p95 = stat.Quantile(0.95, stat.Empirical, durations, nil)
		p99 = stat.Quantile(0.99, stat.Empirical, durations, nil)
		p999 = stat.Quantile(0.999, stat.Empirical, durations, nil)
		smax = floats.Max(durations)
		mean = stat.Mean(durations, nil)
		stdDev = stat.StdDev(durations, nil)
	}
	statistics := metrics.Statistics{
		TotalCount:    totalCount,
		TotalDuration: totalDuration,
		OpsPerSec:     float64(totalCount) / totalDuration,
		Min:           smin,
		P50:           p50,
		P75:           p75,
		P95:           p95,
		P99:           p99,
		P999:          p999,
		Max:           smax,
		Mean:          mean,
		StdDev:        stdDev,
	}
	return statistics
}

func filterByProcessYear(m metrics.DomainMetric) bool {
	_, ok := m.Action.(metrics.ProcessYear)
	return ok
}

func filterByProcessPeriod(m metrics.DomainMetric) bool {
	_, ok := m.Action.(metrics.ProcessPeriod)
	return ok
}
func filterByYearReport(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.YearReport)
		return ok
	}
	return false
}
func filterByPeriodReport(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.PeriodReport)
		return ok
	}
	return false
}
func filterByOpenPeriod(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.OpenPeriod)
		return ok
	}
	return false
}
func filterByClosePeriod(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.ClosePeriod)
		return ok
	}
	return false
}
func filterByCost(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.Cost)
		return ok
	}
	return false
}
func filterBySale(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.Sale)
		return ok
	}
	return false
}
func filterByPurchase(m metrics.DomainMetric) bool {
	op, isOp := m.Action.(metrics.Operation)
	if isOp {
		_, ok := op.Operation.(domain.Purchase)
		return ok
	}
	return false
}

File: rust/src/statistics/save_statistics.rs
use crate::model::domain::Period;
use crate::model::metrics::{Action, DomainMetric, Statistics};
use lazy_static::lazy_static;
use sqlx::{Pool, Postgres};

pub async fn save_statistics(
    pool: Pool<Postgres>,
    benchmark_id: i64,
    metrics: &Vec<DomainMetric>,
    start_year: u32,
    years: u32,
) {

    for rule in PROCESSING_RULES.iter() {
        let action_metrics = metrics.iter().filter(rule.filter_fn).collect::<Vec<_>>();
        let durations = action_metrics
            .iter()
            .map(|m| m.duration)
            .collect::<Vec<_>>();
        let total_stat = calculate_statistics(durations);
        save_stat(pool.clone(), benchmark_id, None, None, rule.action, total_stat).await;

        for year in start_year..(start_year + years) {
            let year_metrics = action_metrics
                .iter()
                .filter(|m| m.year == year)
                .collect::<Vec<_>>();
            let durations = year_metrics.iter().map(|m| m.duration).collect::<Vec<_>>();
            let stat = calculate_statistics(durations);
            save_stat(
                pool.clone(),
                benchmark_id,
                Some(year as i32),
                None,
                rule.action,
                stat,
            )
                .await;

            if rule.calc_periods {
                for month in 1..=12 {
                    let period = Period::new(year as i32, month).month();
                    let durations = year_metrics
                        .iter()
                        .filter(|m| m.period.map(|p| p == period).unwrap_or(false))
                        .map(|m| m.duration)
                        .collect::<Vec<_>>();
                    let stat = calculate_statistics(durations);
                    save_stat(
                        pool.clone(),
                        benchmark_id,
                        Some(year as i32),
                        Some(month),
                        rule.action,
                        stat,
                    )
                        .await;
                }
            }
        }
    }
}


pub struct StatCondition {
    action: &'static str,
    filter_fn: fn(&&DomainMetric) -> bool,
    calc_periods: bool,
}

impl StatCondition {
    fn new(
        action: &'static str,
        filter_fn: fn(&&DomainMetric) -> bool,
        calc_periods: bool,
    ) -> StatCondition {
        StatCondition {
            action,
            filter_fn,
            calc_periods,
        }
    }
}

lazy_static! {
    /// This is an example for using doc comment attributes
    static ref PROCESSING_RULES: Vec<StatCondition> = vec![
        StatCondition::new("PROCESS_YEAR", filter_by_process_year, false),
        StatCondition::new("PROCESS_PERIOD", filter_by_process_period, false),
        StatCondition::new("YEAR_REPORT", filter_by_year_report, true),
        StatCondition::new("PERIOD_REPORT", filter_by_period_report, true),
        StatCondition::new("OPEN_PERIOD", filter_by_open_period, false),
        StatCondition::new("CLOSE_PERIOD", filter_by_close_period, false),
        StatCondition::new("COST", filter_by_cost, true),
        StatCondition::new("SALE", filter_by_sale, true),
        StatCondition::new("PURCHASE", filter_by_purchase, true),
    ];
}


async fn save_stat(
    pool: Pool<Postgres>,
    benchmark_id: i64,
    year: Option<i32>,
    month: Option<i32>,
    action: &str,
    s: Statistics,
) {
    sqlx::query!(
        r#" insert into bm_statistics (
                            benchmark_id, action, year, month, total_count, total_duration, ops_per_sec, 
                            min, p50, p75, p95, p99, p99_9, max, mean, std_dev
                )
              values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16)"#,
        benchmark_id,
        action,
        year,
        month,
        s.total_count as i64,
        s.total_duration,
        s.ops_per_sec,
        s.min,
        s.p50,
        s.p75,
        s.p95,
        s.p99,
        s.p99_9,
        s.max,
        s.mean,
        s.std_dev,
    )
        .execute(&pool)
        .await
        .expect("Cannot insert statistics");
}

fn calculate_statistics(durations: Vec<f64>) -> Statistics {
    let mut percs = inc_stats::Percentiles::new();
    let mut stats = inc_stats::SummStats::new();
    durations.iter().for_each(|d| {
        percs.add(*d);
        stats.add(*d);
    });
    let total_count = percs.count();
    let total_duration = durations.iter().sum::<f64>();
    let p50 = percs.percentile(0.5).unwrap_or(Some(0.0)).unwrap_or(0.0);
    let p75 = percs.percentile(0.75).unwrap_or(Some(0.0)).unwrap_or(0.0);
    let p95 = percs.percentile(0.95).unwrap_or(Some(0.0)).unwrap_or(0.0);
    let p99 = percs.percentile(0.99).unwrap_or(Some(0.0)).unwrap_or(0.0);
    let p99_9 = percs.percentile(0.999).unwrap_or(Some(0.0)).unwrap_or(0.0);
    let statistics = Statistics {
        total_count,
        total_duration,
        ops_per_sec: total_count as f64 / total_duration,
        min: stats.min().unwrap_or(0.0),
        p50,
        p75,
        p95,
        p99,
        p99_9,
        max: stats.max().unwrap_or(0.0),
        mean: stats.mean().unwrap_or(0.0),
        std_dev: stats.standard_deviation().unwrap_or(0.0),
    };
    statistics
}

fn filter_by_process_year(m: &&DomainMetric) -> bool {
    match m.action {
        Action::ProcessYear => true,
        _ => false,
    }
}

fn filter_by_process_period(m: &&DomainMetric) -> bool {
    match m.action {
        Action::ProcessPeriod => true,
        _ => false,
    }
}
fn filter_by_year_report(m: &&DomainMetric) -> bool {
    match m.action {
        Action::YearReport(_) => true,
        _ => false,
    }
}
fn filter_by_period_report(m: &&DomainMetric) -> bool {
    match m.action {
        Action::PeriodReport(_) => true,
        _ => false,
    }
}
fn filter_by_open_period(m: &&DomainMetric) -> bool {
    match m.action {
        Action::OpenPeriod(_) => true,
        _ => false,
    }
}
fn filter_by_close_period(m: &&DomainMetric) -> bool {
    match m.action {
        Action::ClosePeriod(_) => true,
        _ => false,
    }
}
fn filter_by_cost(m: &&DomainMetric) -> bool {
    match m.action {
        Action::Cost(_) => true,
        _ => false,
    }
}
fn filter_by_sale(m: &&DomainMetric) -> bool {
    match m.action {
        Action::Sale(_) => true,
        _ => false,
    }
}
fn filter_by_purchase(m: &&DomainMetric) -> bool {
    match m.action {
        Action::Purchase(_) => true,
        _ => false,
    }
}
